plugins/primus_lisp/semantics/bits.lisp

```diff
@@ -19,13 +19,87 @@
          (logand (msb rm) (lnot (msb rd)))
          (logand (msb rn) (lnot (msb rd)))))
 
-
 (defun overflow (rd rn rm)
   "(overflow RD RN RM) is true if the sum RD = RN + RM results
    in two's complement overflow."
   (logor (logand (msb rn) (msb rm) (lnot (msb rd)))
          (logand (lnot (msb rn)) (lnot (msb rm)) (msb rd))))
 
+(defun highest-set-bit (bitv)
+  "(highest-set-bit bitv) returns the greatest index whose bit is set in bitv.
+   It requires bitv to be non-zero.
+   Translated from ARMv8 ISA pseudocode."
+  (assert-msg (not (is-zero bitv)) "highest-set-bit bitv is zero")  ; at least 1 bit must be set
+  (let ((i (- (word-width bitv) 1)))
+    (while (and (> i 0) (= (select i bitv) 0))
+      (decr i))
+    i))
+
+(defun replicate (bitv n)
+  "(replicate bitv n) returns a bitvector with bitv repeated n times.
+   Translated from ARMv8 ISA pseudocode."
+  (let ((output 0:0))
+    (while (> n 0)
+      (decr n)
+      (set output (concat output bitv)))
+    output))
+
+(defun replicate-to-fill (bitv n)
+  "(replicate-to-fill bitv n) returns the result of repeating bitv
+   to a total of n bits. Requires that n is a multiple of bitv's length.
+   Modified from the bits(N) Replicate(bits(M) x) function from
+   ARMv8 ISA pseudocode."
+  (let ((bitv-length (word-width bitv)))
+    (assert-msg (= 0 (mod n bitv-length)) "replicate-to-fill n not multiple of len(bitv)")
+    (replicate bitv (/ n bitv-length))))
+
+(defun zeros (n)
```

Comment:
This could be done much more efficiently with `(cast-unsigned n 0:1)` the same is true for `ones`, e.g., `(cast-signed n 1:1)`.

Reply:
ok

Reply:
Signed extension doesn't do what we want here; `(cast-signed 4 1:1)` gives `0x9` which is `0b1001`, not `0b1111`. I may just invert `(zeros n)`.

Reply:
This is a bug, `(cast-signed 4 1:1)` must give `0b1111`. I will provide the fix soon!

Reply:
The fix is out #1462

Reply:
yep, using this now